Narrowing types does not consider the @final decorator

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

This is related to #20041. Basically when trying to narrow a `type` object, using `issubclass` makes sense instead of `is` since the narrowing considers that subclasses can be used as replacement. For me however, it makes sense that `issubclass` be the same as `is` when the type is marked `@final` since the subclasses for the type cannot exist. 

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=38818ca2515ec447077ef432725386d7

```python
from typing import final, assert_never, reveal_type

class Types:

    @final
    class A:
        pass

    @final
    class B:
        pass

def pprint(cls: type[Types.A | Types.B]) -> None:
    match cls:
        case Types.A:
            print("A")
            reveal_type(cls)
        case Types.B:
            print("B")
            reveal_type(cls)
        case _:
            reveal_type(cls)
            assert_never(cls)
```

**Expected Behavior**

I expect the types to be narrowed down when using `is` if the type is marked `@final`.

**Actual Behavior**

```
main.py:17: note: Revealed type is "type[__main__.Types.A]"
main.py:20: note: Revealed type is "type[__main__.Types.B]"
main.py:22: note: Revealed type is "type[__main__.Types.A] | type[__main__.Types.B]"
main.py:23: error: Argument 1 to "assert_never" has incompatible type "type[A] | type[B]"; expected "Never"  [arg-type]
```

The type does not narrow, see `main.py:22`

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: mypy 1.19.1 (compiled: yes)
- Mypy command-line flags: 
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.12

<!-- You can freely edit this text, please remove all the lines you believe are unnecessary. -->

--- a/mypy/typeops.py
+++ b/mypy/typeops.py
@@ -1004,6 +1004,10 @@ def is_singleton_identity_type(typ: ProperType) -> bool:
         )
     if isinstance(typ, LiteralType):
         return typ.is_enum_literal() or isinstance(typ.value, bool)
+    if isinstance(typ, TypeType) and isinstance(typ.item, Instance) and typ.item.type.is_final:
+        return True
+    if isinstance(typ, CallableType) and typ.is_type_obj() and typ.type_object().is_final:
+        return True
     return False
 
 

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -3311,3 +3311,44 @@ def type_object(field_type: type[object]) -> str:
             return "bar"
     return ret
 [builtins fixtures/dict.pyi]
+
+
+[case testMatchFinalClass]
+from typing import final, NoReturn
+
+def assert_never(x: NoReturn) -> NoReturn: ...
+
+class Types:
+
+    @final
+    class A:
+        pass
+
+    @final
+    class B:
+        pass
+
+def foo(cls: type[Types.A | Types.B]) -> None:
+    match cls:
+        case Types.A:
+            reveal_type(cls)  # N: Revealed type is "type[__main__.Types.A]"
+        case Types.B:
+            reveal_type(cls)  # N: Revealed type is "type[__main__.Types.B]"
+        case _:
+            assert_never(cls)
+
+
+class TypeTypes:
+    A: type[Types.A] = Types.A
+    B: type[Types.B] = Types.B
+
+
+def bar(cls: type[Types.A | Types.B]) -> None:
+    match cls:
+        case TypeTypes.A:
+            reveal_type(cls)  # N: Revealed type is "type[__main__.Types.A]"
+        case TypeTypes.B:
+            reveal_type(cls)  # N: Revealed type is "type[__main__.Types.B]"
+        case _:
+            assert_never(cls)
+[builtins fixtures/tuple.pyi]

